core/src/main/java/hivemall/tools/map/MapRouletteUDF.java

```diff
@@ -0,0 +1,192 @@
+        if (arguments[0].getCategory() != ObjectInspector.Category.MAP) {
+            throw new UDFArgumentTypeException(0,
+                "Only map type arguments are accepted for the key but " + arguments[0].getTypeName()
+                        + " was passed as parameter 1.");
+        }
+        mapOI = HiveUtils.asMapOI(arguments[0]);
+        ObjectInspector keyOI = mapOI.getMapKeyObjectInspector();
+
+        //judge valueOI is a number
+        valueOI = (PrimitiveObjectInspector) mapOI.getMapValueObjectInspector();
+        switch (valueOI.getTypeName()) {
+            case INT_TYPE_NAME:
+            case DOUBLE_TYPE_NAME:
+            case BIGINT_TYPE_NAME:
+            case FLOAT_TYPE_NAME:
+            case SMALLINT_TYPE_NAME:
+            case TINYINT_TYPE_NAME:
+            case DECIMAL_TYPE_NAME:
+            case STRING_TYPE_NAME:
+                // Pass an empty map or a map full of {null, null} will get string type
+                // An number in string format like "3.5" also support
+                break;
+            default:
+                throw new UDFArgumentException(
+                    "Expected a number but get: " + valueOI.getTypeName());
+        }
+        return keyOI;
+    }
+
+    @Override
+    public Object evaluate(DeferredObject[] arguments) throws HiveException {
+        Map<Object, Double> input = processObjectDoubleMap(arguments[0]);
+        if (input == null) {
+            return null;
+        }
+        // handle empty map
+        if (input.isEmpty()) {
+            return null;
+        }
+        return algorithm(input);
+    }
+
+    /**
+     * Process the data passed by user.
+     * 
+     * @param argument data passed by user
+     * @return If all the value is ,
+     * @throws HiveException If get the wrong weight value like {key = "Wang", value = "Zhang"},
+     *         "Zhang" isn't a number ,this Method will throw exception when
+     *         convertPrimitiveToDouble("Zhang", valueOD)
+     */
+    private Map<Object, Double> processObjectDoubleMap(DeferredObject argument)
+            throws HiveException {
+        // get
+        Map<?, ?> m = mapOI.getMap(argument.get());
+        if (m == null) {
+            return null;
+        }
+        if (m.size() == 0) {
+            return null;
+        }
+        // convert
+        Map<Object, Double> input = new HashMap<>();
+        Double avg = 0.0;
+        for (Map.Entry<?, ?> entry : m.entrySet()) {
+            Object key = entry.getKey();
+            Double value = null;
+            if (entry.getValue() != null) {
+                value = PrimitiveObjectInspectorUtils.convertPrimitiveToDouble(entry.getValue(),
+                    valueOI);
+                if (value < 0) {
+                    throw new UDFArgumentException(entry.getValue() + " < 0");
+                }
+                avg += value;
+            }
+            input.put(key, value);
+        }
+        avg /= m.size();
```

Comment:
`avg /= m.size()` should be revised because average should be computing only with non-null values.

Reply:
`{1, 2, null, null, 3, null}` 

(1+2+3)/3=2
=> `{1, 2, 2, 2, 3, 2}`  

(1+2+3)/3=1.2
=> `{1, 2, 1.2, 1.2, 3, 1.2}`  

consider most of elements are null.